### ISSUE TITLE
Invalid action values when using `visualize_learning`

### ISSUE THREAD

**Issue by philwinder** [david-abel/simple_rl#41]
Hi there. There is a bit of a tricky bug here: https://github.com/david-abel/simple_rl/blob/master/simple_rl/utils/mdp_visualizer.py#L264 when visualising learning.

To replicate, this is a simple test:
```
from simple_rl.agents import QLearningAgent
from simple_rl.tasks import GridWorldMDP

def main():
    mdp = GridWorldMDP(init_loc=(1, 1), goal_locs=[(5, 3)], gamma=1.0)
    ql_agent = QLearningAgent(actions=mdp.get_actions())
    mdp.visualize_learning(ql_agent)

if __name__ == "__main__":
    main()
```
This will only run locally and you will need pygame installed. (`python3 -m pip install pygame==2.0.0.dev3 --pre` was the only version that worked for me on OSX).

I have set the discount factor to 1 (no discounting). Look at the q-values. All boxes should have a value of 1 (the reward of the goal state), since there is no discounting. But note how the expected return continues to increase for ever.

The reason is the code in the visualize learning function. First, it checks for a terminal state in the wrong place. So the q-value at the initial location ends up getting added to pre-goal state. And secondly it does not reset the agent.

In short, can you have a look and make it behave like the `run_single_agent_on_mdp` method?

Basically this line (https://github.com/david-abel/simple_rl/blob/master/simple_rl/utils/mdp_visualizer.py#L228) needs to be moved up to just after the action, and we need an extra `agent.end_of_episode()` call in there too, to reset the agent state.

Thanks!

**Comment by david-abel** [comment]
Ah yes, good catch. If you would like feel free to submit a pull request, otherwise I will make the change suggested (end_of_episode shift) sometime soon, thanks for spotting that!

**Comment by david-abel** [comment]
Should be fixed as of commit [f50b923
](https://github.com/david-abel/simple_rl/commit/f50b92307625ac6f728a946ba72aeed145788430)
